plugins/outputs/postgresql/table_source.go

```diff
@@ -0,0 +1,414 @@
+	}
+	return tsrc
+}
+
+func (tsrc *TableSource) AddMetric(metric telegraf.Metric) {
+	if tsrc.postgresql.TagsAsForeignKeys {
+		tagID := utils.GetTagID(metric)
+		if _, ok := tsrc.tagSets[tagID]; !ok {
+			tsrc.tagSets[tagID] = metric.TagList()
+		}
+	}
+
+	if !tsrc.postgresql.TagsAsJsonb {
+		for _, t := range metric.TagList() {
+			tsrc.tagColumns.Add(ColumnFromTag(t.Key, t.Value))
+		}
+	}
+
+	if !tsrc.postgresql.FieldsAsJsonb {
+		for _, f := range metric.FieldList() {
+			tsrc.fieldColumns.Add(ColumnFromField(f.Key, f.Value))
+		}
+	}
+
+	tsrc.metrics = append(tsrc.metrics, metric)
+}
+
+func (tsrc *TableSource) Name() string {
+	if len(tsrc.metrics) == 0 {
+		return ""
+	}
+	return tsrc.metrics[0].Name()
+}
+
+// Returns the superset of all tags of all metrics.
+func (tsrc *TableSource) TagColumns() []utils.Column {
+	var cols []utils.Column
+
+	if tsrc.postgresql.TagsAsJsonb {
+		cols = append(cols, TagsJSONColumn)
+	} else {
+		cols = append(cols, tsrc.tagColumns.columns...)
+	}
+
+	return cols
+}
+
+// Returns the superset of all fields of all metrics.
+func (tsrc *TableSource) FieldColumns() []utils.Column {
+	return tsrc.fieldColumns.columns
+}
+
+// Returns the full column list, including time, tag id or tags, and fields.
+func (tsrc *TableSource) MetricTableColumns() []utils.Column {
+	cols := []utils.Column{
+		TimeColumn,
+	}
+
+	if tsrc.postgresql.TagsAsForeignKeys {
+		cols = append(cols, TagIDColumn)
+	} else {
+		cols = append(cols, tsrc.TagColumns()...)
+	}
+
+	if tsrc.postgresql.FieldsAsJsonb {
+		cols = append(cols, FieldsJSONColumn)
+	} else {
+		cols = append(cols, tsrc.FieldColumns()...)
+	}
+
+	return cols
+}
+
+func (tsrc *TableSource) TagTableColumns() []utils.Column {
+	cols := []utils.Column{
+		TagIDColumn,
+	}
+
+	cols = append(cols, tsrc.TagColumns()...)
+
+	return cols
+}
+
+func (tsrc *TableSource) ColumnNames() []string {
+	cols := tsrc.MetricTableColumns()
+	names := make([]string, len(cols))
+	for i, col := range cols {
+		names[i] = col.Name
+	}
+	return names
+}
+
+// Drops the specified column.
+// If column is a tag column, any metrics containing the tag will be skipped.
+// If column is a field column, any metrics containing the field will have it omitted.
+func (tsrc *TableSource) DropColumn(col utils.Column) error {
+	switch col.Role {
+	case utils.TagColType:
+		tsrc.dropTagColumn(col)
+	case utils.FieldColType:
+		tsrc.dropFieldColumn(col)
+	case utils.TimeColType, utils.TagsIDColType:
+		return fmt.Errorf("critical column \"%s\"", col.Name)
+	default:
+		return fmt.Errorf("internal error: unknown column \"%s\"", col.Name)
+	}
+	return nil
+}
+
+// Drops the tag column from conversion. Any metrics containing this tag will be skipped.
+func (tsrc *TableSource) dropTagColumn(col utils.Column) {
+	if col.Role != utils.TagColType || tsrc.postgresql.TagsAsJsonb {
+		panic(fmt.Sprintf("Tried to perform an invalid tag drop. This should not have happened. measurement=%s tag=%s", tsrc.Name(), col.Name))
```

Comment:
Is my understanding correct that a user cannot trigger this by misconfiguration but this can only happen on programming errors?

Reply:
Yes, but i should refactor this to not panic. I just missed it.